src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -408,25 +405,6 @@ private void MarkTransitiveOrigin(PackageReference top, PackageReference current
             }
         }
 
-        internal static TransitivePackageReference MergeTransitiveOrigin(PackageReference currentPackage, TransitiveEntry transitiveEntry)
-        {
-            var transitiveOrigins = new SortedSet<PackageReference>(PackageReferenceMergeComparer);
-
-            transitiveEntry?.Keys?.ForEach(key => transitiveOrigins.AddRange(transitiveEntry[key]));
-
-            List<PackageReference> merged = transitiveOrigins
-                .GroupBy(tr => tr.PackageIdentity.Id)
-                .Select(g => g.OrderByDescending(pr => pr.PackageIdentity.Version).First())
```

Comment:
Was exception coming from here?

Reply:
any of the members accessed on this LINQ query clause may be null. Now, there are null checks that guarantee that those values are not null.

Reply:
> Was exception coming from here before your change?

No. Original callstack points to the beginning of the method:

```
NuGet.PackageManagement.VisualStudio.PackageReferenceProject`2.<>c.<.cctor>b__49_0(PackageReference,PackageReference)
System.Collections.Generic.SortedSet`1.AddIfNotPresent(T)
Microsoft.VisualStudio.Services.Common.CollectionsExtensions.AddRange[T,TCollection](TCollection,IEnumerable`1)
Microsoft.VisualStudio.Services.Common.EnumerableExtensions.ForEach[T](IEnumerable`1,Action`1)
NuGet.PackageManagement.VisualStudio.PackageReferenceProject`2.MergeTransitiveOrigin(PackageReference,IDictionary`2)
System.Linq.Enumerable.<>c__DisplayClass7_0`3.<CombineSelectors>b__0(TSource)
System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
System.Collections.Generic.List`1..ctor(IEnumerable`1)
System.Linq.Enumerable.ToList[TSource](IEnumerable`1)
NuGet.PackageManagement.VisualStudio.PackageReferenceProject`2.<GetInstalledAndTransitivePackagesAsync>d__36.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task)
NuGet.VisualStudio.Implementation.Extensibility.NuGetProjectService.<GetInstalledPackagesAsync>d__6.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task)
NuGet.VisualStudio.Implementation.Extensibility.NuGetProjectService.<GetInstalledPackagesAsync>d__5.MoveNext()
```

Reply:
Now, I realized the bug is in the compared. This is also fixed in this PR.

Reply:
My comment is a bit orthogonal to the bug, but the stack trace starts at `NuGet.VisualStudio.Implementation.Extensibility.NuGetProjectService.GetInstalledPackagesAsync`, which is an API we provide to other VS components (Roslyn and Test Explorer are two we know use it), and this API doesn't return transitive origins. While it's nice it caught the bug, `MergeTransitiveOrigin` is doing work that the extensibility API doesn't need, so I hope it's only happening on the first call after the assets file changes. If `MergeTransitiveOrigin` is being called on every `INuGetProjectService.GetInstalledPackagesAsync` is called, then I think we have a potential perf issue.

Reply:
`GetInstalledAndTransitivePackagesAsync()` computes transitive origins once. That was decided on feature development.

Reply:
> `GetInstalledAndTransitivePackagesAsync()` computes transitive origins once.

That doesn't appear to be the implementation, unless I'm reading the code wrong. Note that the permalinks I'm using are for the dev branch, not your PR branch.

It looks like this is the block that gets cached values: https://github.com/NuGet/NuGet.Client/blob/3a55f08732acadee70bbf1eacceca56ae8dacb5e/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs#L121-L143

However, `MergeTransitiveOrigin` is called outside this block: https://github.com/NuGet/NuGet.Client/blob/3a55f08732acadee70bbf1eacceca56ae8dacb5e/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs#L173

In other words, `MergeTransitiveOrigin` is called for every transitive package, every time `GetInstalledPackagesAsync` is called. `MergeTransitiveOrigin` also uses a lot of LINQ, including `OrderByDescending` for every transitive package, meaning this implementation makes a lot of allocations and therefore risks causing a lot of GCs.

So, perhaps the transitive origins are cached, but the additional work that's being performed to "merge" the result looks quite high, which is not needed when anything other than PM UI calls `GetInstalledPackages`, which includes restore and the extensibility APIs we provide to other VS components.

Reply:
We discovered a performance issue when computing transitive origins. Bugfix is here.
https://github.com/NuGet/NuGet.Client/pull/4652/files 

I think it's possible to remove this merge process since PM UI does not support multi-targeting and it would be possible to 'merge' when doing the graph walk. But, first I need the bugfix merged.

Reply:
Update: we still need to merge final results because in multitargeting scenarios, there are multiple graphs with possible different transitive origins.

I will make sure the 'merge' only happens once and then reuse cached results.

Reply:
Update: we still need to merge final results because in multitargeting scenarios, there are multiple graphs with possible different transitive origins.

I will make sure the 'merge' only happens once and then reuse cached results.

Reply:
> > Was exception coming from here before your change?
> 
> No. Original callstack points to the beginning of the method:
> 
> ```
> NuGet.PackageManagement.VisualStudio.PackageReferenceProject`2.<>c.<.cctor>b__49_0(PackageReference,PackageReference)
> System.Collections.Generic.SortedSet`1.AddIfNotPresent(T)
> Microsoft.VisualStudio.Services.Common.CollectionsExtensions.AddRange[T,TCollection](TCollection,IEnumerable`1)
> Microsoft.VisualStudio.Services.Common.EnumerableExtensions.ForEach[T](IEnumerable`1,Action`1)
> NuGet.PackageManagement.VisualStudio.PackageReferenceProject`2.MergeTransitiveOrigin(PackageReference,IDictionary`2)
> System.Linq.Enumerable.<>c__DisplayClass7_0`3.<CombineSelectors>b__0(TSource)
> System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
> System.Collections.Generic.List`1..ctor(IEnumerable`1)
> System.Linq.Enumerable.ToList[TSource](IEnumerable`1)
> NuGet.PackageManagement.VisualStudio.PackageReferenceProject`2.<GetInstalledAndTransitivePackagesAsync>d__36.MoveNext()
> System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
> System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task)
> NuGet.VisualStudio.Implementation.Extensibility.NuGetProjectService.<GetInstalledPackagesAsync>d__6.MoveNext()
> System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
> System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task)
> NuGet.VisualStudio.Implementation.Extensibility.NuGetProjectService.<GetInstalledPackagesAsync>d__5.MoveNext()
> ```

Where did you above call stack? Could you share me a link? I want to see some other details.

Reply:
@erdembayar Shared via private message.

Reply:
@zivkan I will address performance bug in another issue. See https://github.com/NuGet/Client.Engineering/issues/1675

Reply:
Everyone. I will split this PR into smaller ones to address each issue and call stack error separately.